lib/task_listener.js

```diff
@@ -13,166 +13,254 @@ var request = require('superagent-promise');
-        // All content from taskcluster should be a json payload.
-        content = JSON.parse(msg.content);
-        yield self.runTask(content);
-        channel.ack(msg);
-        // Only indicate a completed task (which may trigger an idle state)
-        // after an ack/nack.
-        self.decrementPending();
-      } catch (e) {
-        if (content) {
-          self.runtime.log('task error', {
-            taskId: content.status.taskId,
-            runId: content.runId,
-            message: e.toString(),
-            stack: e.stack,
-            err: e
-          });
-        } else {
-          self.runtime.log('task error', {
-            message: e.toString(),
-            err: e
+    let taskQueue = this.taskQueue;
+    debug(`polling for ${availabileCapacity} tasks`);
+
+    let claims = [];
+    let queues = await taskQueue.getQueues();
+
+    for(let queue of queues) {
+      if (claims >= availabileCapacity) break;
+      // Keep polling queue until enough tasks were claimed or there are no more
+      // tasks in the queue
+      // Move onto the next queue if more tasks are needed but current queue is exhausted
+      // This ensures that as many tasks as possible are consumed from the highest priority
+      // queue
+      while(claims.length < availabileCapacity) {
+        let tasksNeeded = availabileCapacity - claims.length;
+        let tasks = await taskQueue.getTasksFromQueue(queue, availabileCapacity-claims.length);
+        if (!tasks.length) break;
+        let newClaims = await Promise.all(tasks.map(async (task) => {
+          return await this.claimTask(task);
+        }));
+        newClaims.forEach(this.runTask.bind(this));
+        claims = claims.concat(newClaims);
+      }
+    }
+    debug(`Claimed ${claims.length} tasks`);
+  }
+
+  scheduleTaskPoll(nextPoll=5000) {
+    this.pollTimeoutId = setTimeout(() => {
+      async () => {
+        clearTimeout(this.pollTimeoutId);
+
+        try {
+          await this.getTasks();
+        }
+        catch (e) {
+          this.runtime.log('[alert-operator] task retrieval error', {
+              message: e.toString(),
+              err: e
           });
         }
-        var nack = channel.nack(msg, false, false);
-        // Ensure we don't leak pending references.
-        self.decrementPending();
+        this.scheduleTaskPoll();
+      }();
+    }.bind(this), nextPoll);
+  }
+
+  async claimTask(task) {
+    let claim;
+    try {
+      claim = await this.runtime.stats.timeGen(
+        'tasks.time.claim',
+        this.runtime.queue.claimTask(task.taskId, task.runId, this.claimConfig)
+      );
+      await this.taskQueue.deleteTaskFromQueue(task);
+    }
+    catch (e) {
+      // Server error or 401 Authentication errors should stop trying to claim tasks
+      // and not delete the message from the queue
+      if (e && ((500 <= e.statusCode && e.statusCode < 600) || e.statusCode === 401)) {
+        throw e;
       }
```

Comment:
If you don't throw an error here... you should explicitely `return undefined` or `null`... Clearly you don't have a claim.

You might increment a counter for this... I don't know...